tests/python_package_test/test_engine.py

```diff
@@ -48,2591 +49,2655 @@ def categorize(continuous_x):
+                    # test custom eval metrics
+                    feval=(lambda p, d: ('custom_mae', mean_absolute_error(p, d.get_label()), False)),
+                    evals_result=evals_result,
+                    init_model='model.txt')
+    ret = mean_absolute_error(y_test, gbm.predict(X_test))
+    assert ret < 2.0
+    assert evals_result['valid_0']['l1'][-1] == pytest.approx(ret)
+    np.testing.assert_allclose(evals_result['valid_0']['l1'], evals_result['valid_0']['custom_mae'])
+    os.remove(model_name)
+
+
+def test_continue_train_reused_dataset():
+    X, y = load_boston(return_X_y=True)
+    params = {
+        'objective': 'regression',
+        'verbose': -1
+    }
+    lgb_train = lgb.Dataset(X, y, free_raw_data=False)
+    init_gbm = lgb.train(params, lgb_train, num_boost_round=5)
+    init_gbm_2 = lgb.train(params, lgb_train, num_boost_round=5, init_model=init_gbm)
+    init_gbm_3 = lgb.train(params, lgb_train, num_boost_round=5, init_model=init_gbm_2)
+    gbm = lgb.train(params, lgb_train, num_boost_round=5, init_model=init_gbm_3)
+    assert gbm.current_iteration() == 20
+
+
+def test_continue_train_dart():
+    X, y = load_boston(return_X_y=True)
+    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)
+    params = {
+        'boosting_type': 'dart',
+        'objective': 'regression',
+        'metric': 'l1',
+        'verbose': -1
+    }
+    lgb_train = lgb.Dataset(X_train, y_train, free_raw_data=False)
+    lgb_eval = lgb.Dataset(X_test, y_test, reference=lgb_train, free_raw_data=False)
+    init_gbm = lgb.train(params, lgb_train, num_boost_round=50)
+    evals_result = {}
+    gbm = lgb.train(params, lgb_train,
+                    num_boost_round=50,
+                    valid_sets=lgb_eval,
+                    verbose_eval=False,
+                    evals_result=evals_result,
+                    init_model=init_gbm)
+    ret = mean_absolute_error(y_test, gbm.predict(X_test))
+    assert ret < 2.0
+    assert evals_result['valid_0']['l1'][-1] == pytest.approx(ret)
+
+
+def test_continue_train_multiclass():
+    X, y = load_iris(return_X_y=True)
+    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)
+    params = {
+        'objective': 'multiclass',
+        'metric': 'multi_logloss',
+        'num_class': 3,
+        'verbose': -1
+    }
+    lgb_train = lgb.Dataset(X_train, y_train, params=params, free_raw_data=False)
+    lgb_eval = lgb.Dataset(X_test, y_test, reference=lgb_train, params=params, free_raw_data=False)
+    init_gbm = lgb.train(params, lgb_train, num_boost_round=20)
+    evals_result = {}
+    gbm = lgb.train(params, lgb_train,
+                    num_boost_round=30,
+                    valid_sets=lgb_eval,
+                    verbose_eval=False,
+                    evals_result=evals_result,
+                    init_model=init_gbm)
+    ret = multi_logloss(y_test, gbm.predict(X_test))
+    assert ret < 1
```

Comment:
```suggestion
    assert ret < 0.1
```

from https://www.textcompare.org/index.html?id=6008e19c6d59840017ac7a56

![image](https://user-images.githubusercontent.com/7608904/105270916-31780480-5b5c-11eb-9082-351d522c48c2.png)